llama_index/query_engine/jsonalyze_query_engine.py

```diff
@@ -0,0 +1,236 @@
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from llama_index.core.base_query_engine import BaseQueryEngine
+from llama_index.core.response.schema import Response
+from llama_index.prompts import BasePromptTemplate, PromptTemplate
+from llama_index.prompts.default_prompts import DEFAULT_JSONALYZE_PROMPT
+from llama_index.prompts.mixin import PromptDictType, PromptMixinType
+from llama_index.prompts.prompt_type import PromptType
+from llama_index.schema import QueryBundle
+from llama_index.service_context import ServiceContext
+from llama_index.utils import print_text
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RESPONSE_SYNTHESIS_PROMPT_TMPL = (
+    "Given a query, synthesize a response based on SQL query results"
+    " to satisfy the query. Only include details that are relevant to"
+    " the query. If you don't know the answer, then say that.\n"
+    "SQL Query: {sql_query}\n"
+    "Table Schema: {table_schema}\n"
+    "SQL Response: {sql_response}\n"
+    "Query: {query_str}\n"
+    "Response: "
+)
+
+DEFAULT_RESPONSE_SYNTHESIS_PROMPT = PromptTemplate(
+    DEFAULT_RESPONSE_SYNTHESIS_PROMPT_TMPL,
+    prompt_type=PromptType.SQL_RESPONSE_SYNTHESIS,
+)
+
+DEFAULT_TABLE_NAME = "items"
+
+
+def default_jsonalyzer(
+    list_of_dict: List[Dict[str, Any]],
+    query: str,
+    service_context: ServiceContext,
+    table_name: str = DEFAULT_TABLE_NAME,
+    prompt: BasePromptTemplate = DEFAULT_JSONALYZE_PROMPT,
+) -> Tuple[str, Dict[str, Any], List[Dict[str, Any]]]:
+    """Default JSONalyzer that executes a query on a list of dictionaries.
+
+    Args:
+        list_of_dict (List[Dict[str, Any]]): List of dictionaries to query.
+        query (str): The query to execute.
+        prompt (BasePromptTemplate): The prompt to use.
+        service_context (Optional[ServiceContext]): The service context.
+        table_name (str): The table name to use, defaults to DEFAULT_TABLE_NAME.
+
+    Returns:
+        Tuple[str, Dict[str, Any], List[Dict[str, Any]]]: The SQL Query,
+            the Schema, and the Result.
+    """
+    try:
+        import sqlite_utils
+    except ImportError as exc:
+        IMPORT_ERROR_MSG = (
+            "sqlite-utils is needed to use this Query Engine:\n"
+            "pip install sqlite-utils"
+        )
+
+        raise ImportError(IMPORT_ERROR_MSG) from exc
+    # Instantiate in-memory SQLite database
+    db = sqlite_utils.Database(memory=True)
+    try:
+        # Load list of dictionaries into SQLite database
+        db[table_name].insert_all(list_of_dict)
+    except sqlite_utils.db_exceptions.IntegrityError as exc:
+        print_text(f"Error inserting into table {table_name}, expected format:")
+        print_text("[{col1: val1, col2: val2, ...}, ...]")
+        raise ValueError("Invalid list_of_dict") from exc
+
+    # Get the table schema
+    table_schema = db[table_name].columns_dict
+
+    # Get the SQL query with text-to-SQL prompt
+    sql_query = service_context.llm.predict(
```

Comment:
Should we copy the parsing technique from the SQL query engine stuff? Relying on the LLM to print only the SQL query can be a little flakey (it may write extra text around the SQL query)

Reply:
Sure, done!
Thanks!